website/projects/aws/awssync_refactored.py

```diff
@@ -0,0 +1,41 @@
+from __future__ import annotations
+
+import logging
+
+from botocore.exceptions import ClientError
+
+from courses.models import Semester
+
+from projects.aws.awsapitalker import AWSAPITalker
+from projects.aws.awssync_structs import AWSTree
+
+
+class AWSSyncRefactored:
+    """Synchronise with Amazon Web Services."""
+
+    def __init__(self):
+        """Create an AWSSync instance."""
+        self.api_talker = AWSAPITalker()
+
+        self.logger = logging.getLogger("django.aws")
+        self.logger.setLevel(logging.DEBUG)
+
+    def create_course_ou(self, tree: AWSTree) -> str:
+        """Create organizational unit under root with name of current semester."""
+        root_id = tree.ou_id
+        course_ou_name = str(Semester.objects.get_or_create_current_semester())
+        course_ou_id = next((ou.ou_id for ou in tree.iterations if ou.name == course_ou_name), None)
+
+        if not course_ou_id:
+            course_ou = self.api_talker.create_organizational_unit(root_id, course_ou_name)
+            course_ou_id = course_ou["OrganizationalUnit"]["Id"]
+
+        return course_ou_id
+
+    def attach_policy(self, target_id: str, policy_id: str) -> None:
+        """Attach policy to target resource."""
+        try:
```

Comment:
> Any client errors except for the "DuplicatePolicyAttachmentException" are now caught and handled by doing nothing with it. This means they also won't throw an exception in the main try/except block in the pipeline. Is "DuplicatePolicyAttachmentException" the only exception that should cause abortion of the pipeline?

It is the opposite of what you write. A "hard fail" signal is sent to the pipeline (by re-reraising) for all `ClientError` exceptions because it indicates something went wrong with attaching the policy to a target resource, except if the policy was already attached before (indicated by AWS error code `DuplicatePolicyAttachmentException`). I have to catch the `ClientError` to determine if it was already bound based on the returned AWS error.

The alternative would have been to manually check. I had included such an alternative function previously in the branch if you remember. The main difference is in manually checking versus letting AWS check if a policy is already attached to a resource.

```Python
    # Alternative implementation of the above. Requires some manual checks instead of letting AWS check.
    def attach_scp_policy(self, target_id: str, policy_id: str) -> None:  # pragma: no cover
        """Attach policy to target resource."""
        client = boto3.client("organizations")  # TODO: Add list_policies_for_target to AWSAPITalker.
        policies = client.list_policies_for_target(TargetId=target_id, Filter="SERVICE_CONTROL_POLICY")["Policies"]
        policy_ids = [policy["Id"] for policy in policies]

        if policy_id not in policy_ids:
            self.api_talker.attach_policy(target_id, policy_id)
```

Reply:
I misread the if statement check. So this solution is good.